walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -49,93 +58,120 @@
     Event endOfDay = new Event(endOfDayTitle, TimeRange.fromStartDuration(TimeRange.END_OF_DAY + 1, 0), allAttendees);
     Event startOfDay = new Event(startOfDayTitle, TimeRange.fromStartDuration(TimeRange.START_OF_DAY, 0), allAttendees);
 
-    ArrayList<Event> allEvents = new ArrayList<Event>(events);
+    LinkedList<Event> sortedEvents = new LinkedList<Event>(events);
 
-    allEvents.add(endOfDay);
-    allEvents.add(startOfDay);
+    sortedEvents.add(endOfDay);
+    sortedEvents.add(startOfDay);
       
-    allEvents.sort(Event.EVENT_COMPARATOR);
-    
-    ArrayList<TimeRange> available = availableTimesFromSortedEvents(mandatoryAttendees, allEvents, request);
+    sortedEvents.sort(Event.EVENT_COMPARATOR);
+
+    LinkedList<Event> applicableEventsOnly = removeEventsWithNoAttendees(mandatoryAttendees, sortedEvents);
+    LinkedList<Event> mandatoryNonOverlapping = sortedEventsIntoNonOverlappingSortedEvents(mandatoryAttendees, applicableEventsOnly);
+    LinkedList<TimeRange> available = availableTimesFromSortedNonOverlappingEvents(mandatoryNonOverlapping, request);
 
     if (optionalAttendees.size() == 0) {
       return available;
     }
 
-    ArrayList<TimeRange> availableWithOptional = availableTimesFromSortedEvents(optionalAttendees, allEvents, request);
+    LinkedList<Event> applicableEventsOnlyOfOptional = removeEventsWithNoAttendees(optionalAttendees, sortedEvents);
+    LinkedList<Event> optionalNonOverlapping = sortedEventsIntoNonOverlappingSortedEvents(optionalAttendees, applicableEventsOnlyOfOptional);
+    LinkedList<TimeRange> availableWithOptional = availableTimesFromSortedNonOverlappingEvents(optionalNonOverlapping, request);
     
     if (available.size() == 0) {
       return availableWithOptional;
     }
     
-    ArrayList<TimeRange> finalTimeRanges = new ArrayList<>();
+    LinkedList<TimeRange> finalTimeRanges = mergeTimeRangesThatIntersect(availableWithOptional, available);
+    return finalTimeRanges;
+  }
+   
+  /*
+   * Function to merge time ranges that intersect between two lists of time ranges
+   * 
+   * Runtime: O(n*logm) where 'n' is the length of optional and 'm' is the length of mandatory
+   * 
+   * @param optional    A list of time ranges in which for every one we'll search for an intersection
+   * @param mandatory   A list of time ranges to search within
+   * 
+   * return   A list of time ranges where all ranges are within both input lists
+   * 
+  */
+  private LinkedList<TimeRange> mergeTimeRangesThatIntersect(LinkedList<TimeRange> optional, LinkedList<TimeRange> mandatory) {
```

Comment:
Yes they do. I need a binary search because for every optional time range I need to see if it overlaps with any mandatory time range. So I search any mandatory time range which both starts before or equal and ends after or equal to the optional time range.

Reply:
Hmm, you might have an opportunity here to optimize and do something similar to computing the union of two sorted sets, which is `O(N + M)`: https://www.geeksforgeeks.org/union-and-intersection-of-two-sorted-arrays-2/.

But this algorithm definitely works, it's totally fine not to have it fully optimized right now.

Reply:
Ohhhh how did I not think of that. Yeah makes sense